File: cpp11/CountDownLatch.cpp
#include "CountDownLatch.h"

CountDownLatch::CountDownLatch(int count)
    : count_(count) {}

void CountDownLatch::wait() {
  //MutexLockGuard lock(mutex_);
  std::unique_lock<std::mutex> lock(mutex_);
  //while (count_ > 0) condition_.wait();
  condition_.wait(lock, [=] {return count_ <= 0;});
}

void CountDownLatch::countDown() {
  //MutexLockGuard lock(mutex_);
  std::unique_lock<std::mutex> lock(mutex_);
  --count_;
  if (count_ == 0) condition_.notify_all();
}

File: go_synchronous_mechanism/src/LogModule/CountDownLatch.go
package LogModule

import (
	"sync"
)

type CountDownLatch struct {
	mutex *sync.Mutex
	cond *sync.Cond
	count int
}

func NewCountDownLatch(count int) *CountDownLatch {
	mutex := &sync.Mutex{}
	cond := sync.NewCond(mutex)
	return &CountDownLatch{mutex: mutex, cond: cond, count: count}
}

func (c *CountDownLatch) Wait() {
	c.mutex.Lock()
	defer c.mutex.Unlock()
	for c.count > 0 {
		c.cond.Wait()
	}
}

func (c *CountDownLatch) CountDown() {
	c.mutex.Lock()
	defer c.mutex.Unlock()
	c.count -= 1
	if c.count == 0 {
		c.cond.Broadcast()
	}
}